### ISSUE TITLE
Rspamd web-interface configuration not saving

### ISSUE THREAD

**Issue by 1n5aN1aC** [hardware/mailserver#179]
Changes made via the rspamd web-interface (specifically the Configuration tab) do not seem to actually be saved.

For example, if I go to the WebUI, then go to the Configuration tab, and change either the "rewrite subject" or "spam" values, they appear to unreliably both be reset back to 8.  There's some funny business going on where I can sometimes get them to change visually, but upon a full reload of the page, both are back to 8.

This is odd, because the default for "spam" should be 15, and I don't think I've ever changed it, while I have changed some of the other entries, but now I cannot get it to go above 8.  :(

Anyways, if you have any ideas, or think I should go to rspamd themselves let me know.  But it looks like you don't use the normal config file at "/etc/rspamd/local.d/metrics.conf" in the docker?  As those are not changed from their default values.

**Comment by hardware** [comment]
> Changes made via the rspamd web-interface (specifically the Configuration tab) do not seem to actually be saved.

You can ask on the rspamd mailing list how it works. Both DBDIR and Redis database are persistent across reboots.

> This is odd, because the default for "spam" should be 15

What action ? (reject, soft_reject, rewrite_subject, add_header, greylist)

> Anyways, if you have any ideas

You can overwrite with a ro volume on `/etc/rspamd/local.d/metrics.conf``.

**Comment by 1n5aN1aC** [comment]
> What action ? (reject, soft_reject, rewrite_subject, add_header, greylist)

Well, that's what I'm confused by.  The web interface lists "Greylist", "Probably Spam", "Rewrite Subject", and "Spam".  I'm assuming that:

- Greylist = Greylist
- Probably Spam = Add_header
- Rewrite Subject = Rewrite_Subject
- and Spam = Reject

I can confirm that when the web interface shows "8" as the setting for "Spam", it is rejecting emails that achieve an 8.  Not 100% sure which type of reject, but they never hit a mailbox.

>You can overwrite with a read-only volume the /etc/rspamd/local.d/metrics.conf file.

This is my main question I have for you.  If I go into the docker with `docker exec -ti mailserver bash`, and access that specific file, it shows all the defaults:

    actions {
      reject = 20;
      soft_reject = 15;
      rewrite_subject = 8;
      add_header = 6;
      greylist = 4;
    }

Yet these are somehow not being used, as I've seen several emails in the rspamd history that are being blocked with spam levels of 8.something, so it appears that rspamd is not using that config file.  Do you have any idea why that might be?

**Comment by hardware** [comment]
> I can confirm that when the web interface shows "8" as the setting for "Spam", it is rejecting emails that achieve an 8. Not 100% sure which type of reject, but they never hit a mailbox.

I can confirm this too. My WebUI shows 20 for spam level rejection, and at this level they never hit a mailbox.

> as I've seen several emails in the rspamd history that are being blocked with spam levels of 8.something, so it appears that rspamd is not using that config file.

Blocked with spam level of 8 with a value of 20 in WebUI and metrics.conf ? That's not possible.

**Comment by 1n5aN1aC** [comment]
Sorry if I was not clear on the discrepancies I was trying to explain.

What I was trying to explain is that `/etc/rspamd/local.d/metrics.conf` shows (as far as I know) the defaults.  Those being as I posted above.  20 for reject.

The problem is that the WebUI is showing 8, and 8 is what is being applied-  as if the config file is being ignored completely, and I have no idea why.

**Comment by hardware** [comment]
Ok I see. Try this :

```
docker exec mailserver rspamadm configdump > /tmp/rspamd-config-dump.txt
```

Open the dump file and check actions values in the metric block.

**Comment by 1n5aN1aC** [comment]
This is very odd.....  Here is my output:

    lua = "/usr/share/rspamd/rules/rspamd.lua";
    metric {
        name = "default";
        actions {
            reject = 20;
            add_header = 6;
            greylist = 4;
            soft_reject = 15;
            rewrite_subject = 8;
        }
    ...and much more, but no duplicate actions...

Yet....  The configuration tab shows 8, and any changes I try to make are reverted upon reload.  (I'm assuming never saved):
![](https://i.imgur.com/DB9clB4.png)

And it HAS been rejecting with a metric of 8:
![](https://i.imgur.com/0HTx1pM.png)

**Comment by hardware** [comment]
You can ask on the rspamd mailing list how it works because i have not idea what is going on in this case.

**Comment by hardware** [comment]
@1n5aN1aC can you try the latest build on master (`1.1-latest` branch) ?

**Comment by 1n5aN1aC** [comment]
@hardware I had hope this would resolve the issue, but it seems to persist.

I'll do more testing tomorrow, and hopefully can figure out what's going on.....

**Comment by 1n5aN1aC** [comment]
Sorry for the delay!

Updated mailserver again today, (hardware/mailserver:1.1-latest) and I can confirm, same result.

the file in `/etc/rspamd/local.d/metrics.conf` has indeed had the line in question removed, but the webGUI still shows 8 for both "rewrite subject" and "spam."  Furthermore, I can confirm that 8 is indeed being applied.

I'm not able to easily spin up another image, but I'll try to do so when I get home.

**Comment by hardware** [comment]
@1n5aN1aC can you try the latest build on master ? I updated rspamd to [1.6.5](https://github.com/vstakhov/rspamd/releases/tag/1.6.5) version.

**Comment by 1n5aN1aC** [comment]
Unfortunately, that still didn't make any difference.  It's still stuck at 8, and won't let me change it any higher.

Sorry that I have still not been able to spin up another image with new configs, rather than the current config directory.  Sorry, I'm hoping to do that soon, but just haven't had the time yet.  :/

**Comment by fatalbanana** [comment]
Pretty sure rescoring actions in web UI only works if there are exactly *3* actions scored: most likely `greylist`, `add header` and `reject` (no `rewrite subject` here). You'll notice it only posts 3 values when you click 'Save'. Whether it will definitely work given `rewrite subject` is removed from metric I'm not sure.

**Comment by 1n5aN1aC** [comment]
Apparently, I haven't been getting Github notifications, which I didn't realize until I never got a very important email I was expecting, only to find it had been hard rejected by rpamd.  (not rspamd's fault, the email was horrible, didn't even have a mx record, among numerous other issues.)  I remembered this topic, only to find Github hadn't been sending me notification emails.  Got that resolved, now it's back to this issue.

Anyways, what @fatalbanana says makes complete sense, as it does only send 3 values when you click post.

I just now tried updating the container to latest (2b1de9436c91), No change.  
I just now removed the file `rspamd/rspamd_dynamic` in my docker config directory, restarted the container, and voila!  It works.

Unfortunately, I don't know exactly what was in there, but I would expect there was an `actions {` definition that was added wrongly from the web interface, and breaking it.
